<div style="border:solid green 2px; padding: 20px"> <h1 style="color:green; margin-bottom:20px">Reviewer's comment v1</h1>

Hello Amanda!

I'm happy to review your project today.  🙌

You can find my comments under the heading **«Review»**. I will categorize my comments in green, blue or red boxes like this:

<div class="alert alert-success">
    <b>Success:</b> if everything is done successfully
</div>
<div class="alert alert-warning">
    <b>Remarks:</b> if I can give some recommendations or ways to improve the project
</div>
<div class="alert alert-danger">
    <b>Needs fixing:</b> if the block requires some corrections. Work can't be accepted with the red comments
</div>

Please don't remove my comments. If you have any questions, don't hesitate to respond to my comments in a different section.
<div class="alert alert-info"> <b>Student comments: </div>    


<div style="border:solid green 2px; padding: 20px">
<b>Reviewer's comment v1:</b>
    
<b>Overall Feedback</b> 
    
Overall well done! I can see that a lot of effort has been made! Your project looks very good and you accomplished impressive results. There are no critical issues, so your project has been accepted. 

Please keep up great work and don't hesitate to use this project as a reference in your future sprints.

Good luck on the next project 🍀

    
</div>

## Basic Python - Project <a id='intro'></a>

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
It is always helpful for the reader to have additional information about project tasks. It gives an overview of what you are going to achieve in this project.


## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

You'll need `pandas`, so import it.

In [1]:
import pandas as pd # importing pandas


Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [2]:
# reading the files and storing them to df
df = pd.read_csv('/datasets/movies_and_shows.csv')

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Well done! Required library has been imported.

Print the first 10 table rows:

In [3]:
df.head(10)# obtaining the first 10 rows from the df table
# hint: you can use head() and tail() in Jupyter Notebook without wrapping them into print()


name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

Obtain the general information about the table with one command:

In [4]:
df.describe()# obtaining general information about the data in df


release Year    imdb sc0re    imdb v0tes
count  85579.000000  80970.000000  8.085300e+04
mean    2015.879994      6.425877  5.978271e+04
std        7.724668      1.122655  1.846287e+05
min     1954.000000      1.500000  5.000000e+00
25%     2015.000000      5.700000  1.266000e+03
50%     2018.000000      6.500000  5.448000e+03
75%     2021.000000      7.200000  3.360900e+04
max     2022.000000      9.500000  2.294231e+06

The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Great data overview and correct conclusions. 

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

In [5]:
df.columns.tolist()# the list of column names in the df table


['   name',
 'Character',
 'r0le',
 'TITLE',
 '  Type',
 'release Year',
 'genres',
 'imdb sc0re',
 'imdb v0tes']

Change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [6]:
df = df.rename(
    columns={
        '   name':'name',
        'Character': 'character',
        'r0le': 'role',
        'TITLE':'title',
        '  Type': 'type',
        'release Year': 'release_year',
        'imdb sc0re': 'imdb_score',
        'imdb v0tes': 'imdb_votes',
        
        
        }
    )# renaming columns


Check the result. Print the names of the columns once more:

In [7]:
df.columns.tolist()# checking result: the list of column names


['name',
 'character',
 'role',
 'title',
 'type',
 'release_year',
 'genres',
 'imdb_score',
 'imdb_votes']

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
It is great that you correctly applied `rename` method. 

### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [8]:
df.isna().sum()# calculating missing values


name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64

<div class="alert alert-success">
<b>Reviewer's comment v1</b>
    
Great that you selected `isna()` method to find missing values! 

It is also sometimes helpful to check not only the total amount of missing values in each column but also look at the percentage of missing values. It helps to understand the overall impact. You can check percentage using, for example, this code:

`df.isnull().sum()/len(df)`

Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` and `'imdb_votes'` columns.

In [9]:
df = df.dropna(subset = ['imdb_score', 'imdb_votes'])# dropping rows where columns with title, scores and votes have missing values


Make sure the table doesn't contain any more missing values. Count the missing values again.

In [10]:
df.isna().sum()# counting missing values


name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Everything is correct here!

### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [11]:
df.duplicated().sum()# counting duplicate rows


6994

Review the duplicate rows to determine if removing them would distort our dataset.

In [12]:
df[df.duplicated()]# Produce table with duplicates (with original rows included) and review last 5 rows


name                  character      role  \
7561         Philip Greene  Baseball Fan (uncredited)     ACTOR   
14512             Dan Levy                   Reporter     ACTOR   
18952     Nicolas Le Nev??                    unknown  DIRECTOR   
22456    John Iii Franklin                    Himself     ACTOR   
29557         Claudio Roca                   Nicol?­s     ACTOR   
...                    ...                        ...       ...   
85569       Jessica Cediel            Liliana Navarro     ACTOR   
85570  Javier Gardeaz?­bal   Agust??n "Peluca" Ort??z     ACTOR   
85571        Carla Giraldo             Valery Reinoso     ACTOR   
85572  Ana Mar??a S?­nchez                    Lourdes     ACTOR   
85577         Isabel Gaona                     Cacica     ACTOR   

                                 title   type  release_year  \
7561                   How Do You Know  MOVIE          2010   
14512  A Very Harold & Kumar Christmas  MOVIE          2011   
18952                       Sammy & Co   SHOW          2014   
22456                    Last Chance U   SHOW          2016   
29557                   Narcos: Mexico   SHOW          2018   
...                                ...    ...           ...   
85569                          Lokillo  MOVIE          2021   
85570                          Lokillo  MOVIE          2021   
85571                          Lokillo  MOVIE          2021   
85572                          Lokillo  MOVIE          2021   
85577                          Lokillo  MOVIE          2021   

                                 genres  imdb_score  imdb_votes  
7561     ['comedy', 'drama', 'romance']         5.4     50383.0  
14512  ['comedy', 'fantasy', 'romance']         6.2     69562.0  
18952         ['animation', 'european']         5.7        31.0  
22456        ['documentation', 'sport']         8.4      6897.0  
29557                ['drama', 'crime']         8.4     82042.0  
...                                 ...         ...         ...  
85569                        ['comedy']         3.8        68.0  
85570                        ['comedy']         3.8        68.0  
85571                        ['comedy']         3.8        68.0  
85572                        ['comedy']         3.8        68.0  
85577                        ['comedy']         3.8        68.0  

[6994 rows x 9 columns]

<div class="alert alert-warning">
<b>Reviewer's comment v1</b>
    
Note that it is also required additionally to re-create the indexes in your dataframe. To achieve that, you can use `reset_index()`.
    
Without that, the index column will not be ordinal anymore, as `drop_duplicates()` deleted some lines, hence the dataframe becomes less informative.
    
    
You can read about it here: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html
    
And parameter `drop=True` could be helpful to rewrite existent index instead of creating a second ordinal column.

There are two clear duplicates in the printed rows. We can safely remove them.
Call the `pandas` method for getting rid of duplicate rows:

In [13]:
df = df.drop_duplicates()# removing duplicate rows


Check for duplicate rows once more to make sure you have removed all of them:

In [14]:
df.duplicated().sum()# checking for duplicates


0

<div class="alert alert-success">
<b>Reviewer's comment v1</b>
    
You correctly found and deleted all duplicates from the data.

Now get rid of implicit duplicates in the `'type'` column. For example, the string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique `'type'` names, sorted in alphabetical order. To do so:
* Retrieve the intended dataframe column 
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [15]:
df['type'].sort_values().unique()# viewing unique type names


array(['MOVIE', 'SHOW', 'movies', 'shows', 'the movie', 'tv', 'tv series',
       'tv show', 'tv shows'], dtype=object)

Look through the list to find implicit duplicates of `'show'` (`'movie'` duplicates will be ignored since the assumption is about shows). These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, declare the function `replace_wrong_show()` with two parameters: 
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function should correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

In [16]:
    def replace_wrong_show(wrong_shows_list, correct_show): 
          df['type'] = df['type'].replace(wrong_shows_list, correct_show)# function for replacing implicit duplicates


Call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `SHOW`:

In [17]:

wrong_shows = ['shows', 'SHOW', 'tv show', 'tv shows', 'tv series', 'tv']

correct_show = 'show'

replace_wrong_show(wrong_shows, correct_show)# removing implicit duplicates

Make sure the duplicate names are removed. Print the list of unique values from the `'type'` column:

In [18]:
df['type'].unique()# viewing unique genre names


array(['MOVIE', 'the movie', 'show', 'movies'], dtype=object)

<div class="alert alert-success">
<b>Reviewer's comment v1</b>
    
Your function works perfectly fine.
    
Btw you can also just pass values in your function:
    
`replace_wrong_genres(['shows', 'SHOW', 'tv show', 'tv shows', 'tv series', 'tv'], 'SHOW')`
 

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

In [19]:
newer_1999 = df[(df['release_year'] >= 1999)]# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)
# give the slice of dataframe new name


In [20]:
newer_1999 = df[(df['release_year'] >= 1999) & (df['type'] == 'show')].copy()# repeat conditional indexing so df has only shows (movies are removed as result)


<div class="alert alert-success">
<b>Reviewer's comment v1</b>

You've filterd the data required data correctly. 


Additionally there are other methods that could be helpful to filter your data, for example:`query()`.

    
`df = df.query('type == "SHOW"')`
    
You can read about them additionally here: 
https://towardsdatascience.com/10-examples-that-will-make-you-use-pandas-query-function-more-often-a8fb3e9361cb
https://towardsdatascience.com/how-to-use-loc-and-iloc-for-selecting-data-in-pandas-bd09cb4c3d79    

The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

In [21]:
newer_1999.loc[:,'imdb_score'] = newer_1999['imdb_score'].round()# rounding column with scores

newer_1999.tail()#checking the outcome with tail()


name character      role      title  type  release_year  \
85433     Maneerat Kam-Uan        Ae     ACTOR  Let's Eat  show          2021   
85434    Rudklao Amratisha   unknown     ACTOR  Let's Eat  show          2021   
85435      Jaturong Mokjok   unknown     ACTOR  Let's Eat  show          2021   
85436     Pisamai Wilaisak   unknown     ACTOR  Let's Eat  show          2021   
85437  Sarawut Wichiensarn   unknown  DIRECTOR  Let's Eat  show          2021   

                    genres  imdb_score  imdb_votes  
85433  ['drama', 'comedy']         8.0         5.0  
85434  ['drama', 'comedy']         8.0         5.0  
85435  ['drama', 'comedy']         8.0         5.0  
85436  ['drama', 'comedy']         8.0         5.0  
85437  ['drama', 'comedy']         8.0         5.0

It is now time to identify outliers based on the number of votes.

In [22]:
newer_1999.groupby('imdb_score')['imdb_votes'].count()# Use groupby() for scores and count all unique values in each group, print the result


imdb_score
2.0       24
3.0       27
4.0      180
5.0      592
6.0     2494
7.0     4706
8.0     4842
9.0      557
10.0       8
Name: imdb_votes, dtype: int64

Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [23]:
filtered_df = newer_1999[(newer_1999['imdb_score']>= 4) & (newer_1999['imdb_score'] <= 9)]# filter dataframe using two conditions (scores to be in the range 4-9)

average_votes = filtered_df.groupby('imdb_score')['imdb_votes'].mean().reset_index()# group scores and corresponding average number of votes, reset index and print the result


Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [24]:
average_votes['imdb_votes'] = average_votes['imdb_votes'].round()# round column with averages

average_votes.columns = ['imdb_score', 'average_votes']# rename columns

average_votes = average_votes.sort_values(by='average_votes', ascending=False)

average_votes# print dataframe in descending order


imdb_score  average_votes
5         9.0       126904.0
4         8.0        30299.0
3         7.0         8727.0
0         4.0         5278.0
2         6.0         3482.0
1         5.0         3144.0

<div class="alert alert-success">
<b>Reviewer's comment v1</b>

Everything is correct here. 

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
 
Overall you did a great research.
    
A small tip regarding overall conclusion in the project: 
    It represents the overall work progress that you achieved. On a real project, this is probably the only thing the business will read. Therefore, it is crucial to indicate in a structured way all conclusions that you made on each step in the project.

For example:

- Replaced missing values in the following data with the following method.
- Replaced data types in the following columns.
- etc.
- We observe that ... factors impact ... 
- My analysis show ...
- I can recommend the following next steps / activities ...

It is also important to provide explanations and interpretations that will be interesting for business based on your analysis.

</div>